pkg/tfgen/docs.go

```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
 		contract.IgnoreError(err)
 	}
 
-	for _, section := range splitGroupLines(docs, "## ") {
-		if len(section) == 0 {
-			continue
+	// Find code fences in the presented doc
+	type tfBlockIndex struct {
+		start       int
+		end         int
+		headerStart int // a value of -1 indicates there's no header
+	}
+
+	var inCodeBlock bool
+	codeFence := "```"
+	h2 := "##"
+	h3 := "###"
+	var codeIndices []tfBlockIndex
+	var currentBlock tfBlockIndex
+	currentBlock.headerStart = -1
+
+	for i := 0; i < (len(docs) - len(codeFence)); i++ {
+
+		if inCodeBlock == true {
+			//Our TF code block is over right when we see the closing code fence.
+			if docs[i:i+len(codeFence)] == codeFence {
+				currentBlock.end = i
+				codeIndices = append(codeIndices, currentBlock)
+				// reset
+				currentBlock.start = 0
+				currentBlock.end = 0
+				inCodeBlock = false
+			}
+		} else {
+			// Keep track of header locations. These should never be inside code blocks.
+			if docs[i:i+len(h2)] == h2 || docs[i:i+len(h3)] == h3 {
+				// We need to make sure we don't reread an H3 as H2 on the next pass
+				if !(currentBlock.headerStart+1 == i) {
+					currentBlock.headerStart = i
+				}
+			}
+			// if we aren't in a code block, a set of code fences signals the beginning of a code block.
+			if docs[i:i+len(codeFence)] == codeFence {
+				inCodeBlock = true
+				currentBlock.start = i
+			}
 		}
+	}
 
-		isImportSection := false
-		header, wroteHeader := section[0], false
-		isFrontMatter, isExampleUsage := !strings.HasPrefix(header, "## "), header == "## Example Usage"
+	// Now we know where the headers and code fences are.
+	textStart := 0
```

Comment:
There's quite a few interactions and context one needs to keep when reading here, would this also make sense to be split into a few functions?